### DISCUSSION TITLE
Some questions about generics

### DISCUSSION THREAD

**Discussion by Luosuu** [google/comprehensive-rust#124] [Q&A] [answered]
https://google.github.io/comprehensive-rust/generics/impl-trait.html

Does this mean the return type of `get_x` function is `impl Display`? I am a bit confused. 

BTW, I think the code in this page can be more understandable like

```rust
use std::fmt::Display;

fn get_x(name: impl Display) -> impl Display {
    format!("Hello {name}")
}

fn main() {
    let x = get_x("foo");
    let y = get_x(123);
    println!("{x}");
    println!("{y}")
}
```
This can better show the idea of 

> `impl Trait` allows you to work with types which you cannot name.

**Comment by mgeisler** [comment, marked as answer]
Hi @Luosuu,

Yeah, the instructor should highlight that `get_x` can be called on both integers and strings! Let us add that to the speaker notes...

As for the return type, you're right that it's `impl Display`. The return type is chosen by the person who _writes_ `get_x`. The _caller_ can only rely on `Display` being implemented, nothing more.

This is different from the generic in the argument position: the type of `name` is chosen by the _caller_ of `get_x`. The person who write the function cannot know anything about the type, except that it implements `Display`. This allows `name` to be used with `format!` as shown in the example.

Does that help a bit? Generics is a complex topic and I need to expand the slides with significant amounts of speaker notes to fully explain this :smile:

**Reply by Luosuu** [reply]
Thank you very much!
